src/compiler/program.ts

```diff
@@ -498,15 +515,25 @@ namespace ts {
                         return false;
                     }
 
-                    // check imports
+                    // check imports and module augmentations
                     collectExternalModuleReferences(newSourceFile);
                     if (!arrayIsEqualTo(oldSourceFile.imports, newSourceFile.imports, moduleNameIsEqualTo)) {
                         // imports has changed
                         return false;
                     }
+                    if (!arrayIsEqualTo(oldSourceFile.moduleAugmentations, newSourceFile.moduleAugmentations, moduleNameIsEqualTo)) {
+                        // moduleAugmentations has changed
+                        return false;
+                    }
 
                     if (resolveModuleNamesWorker) {
-                        const moduleNames = map(newSourceFile.imports, name => name.text);
+                        const moduleNames: string[] = [];
+                        for (const moduleName of newSourceFile.imports) {
+                            moduleNames.push(moduleName.text);
+                        }
+                        for (const moduleName of newSourceFile.moduleAugmentations) {
+                            moduleNames.push(moduleName.text);
+                        }

```

Comment:
Just FYI, the `concatenate` function treats `undefined` the same as an empty array and only allocates a new array if both arguments are defined. Otherwise it just returns the one defined argument. Since it'll be rare for both to be defined here, it's fine to use this function instead of the imperative loops.